pkg/boot/netboot/ipxe/ipxe.go

```diff
@@ -86,7 +87,42 @@ func (c *parser) getFile(surl string) (io.ReaderAt, error) {
 	if err != nil {
 		return nil, fmt.Errorf("could not parse URL %q: %v", surl, err)
 	}
-	return c.schemes.LazyFetch(u)
+	// return c.schemes.LazyFetch(u)
```

Comment:
the big question I have: we're doing heap avoidance but copying a file to tmpfs. But it's all ram at that point. 

If you could mmap, and then read into the file directly, avoiding the bounce buffers you are allocating, ISTM this would be faster. It's what I've done anyway. 

Overall, if you are copying from ram to ram, I'd consider that as a problem and start there. But I know you've done a lot with mmap, so maybe there is something here I just don't know.

Reply:
You are spot on, Ron. I think mmap is also a distraction -- you can use small buffers to stream reading from or writing to a file in tmpfs.

Reply:
The memory number that really matters, the one that you need to track with these changes, is ultimately the amount of physical memory used by a full build. (You can do that with one of the VMs we have internally or the qemu stuff here.)

Reply:
This code is one of the approaches that I experimented with, and eventually went with. I think the big question asked you is **mis-led** by this code itself. 

Let me attempt to explain how I arrived at this version, rather than others.

**The problem**: before this change, the `c.schemes.LazyFetch(u)` returns an `io.ReaderAt` (with an in heap bytes slice to support seek-able read). That in-heap cache[1] is the main thing I am trying to get ride of. 

**Approaches tried**: 

`breaking interface change`: I tried change LinuxImage's reference to `kernel` and `initrd` to be `io.Reader`, from current `io.ReaderAt`. Then we don't need the cache to support seek-able read, `io.ReaderAt`. It did not go well, as this is breaking interface change, and seek-able read is be relied on pervasively in other boot paths. It is a workable approach. But, eventually, I gave up, as the volume of code turned out to be a lot, and more importantly, I have to read it once into a giant bytes slice, and then have all code that previously reads `io.ReaderAt` to instead read from the pre-read bytes slice. You now probably can tell that, this effectively created another `"copy"` of the image in heap, that we are trying to get ride of in the first place, as well ( well, at least transiently while the bytes slice are still referenced. 

`Loads contents into file into tmpfs`: Reading http body `io.Reader`, and writting it directly into tmpfs file help avoid maintaining big slice in heap. In addition, with this, I can skip calling `curl.http.Fetch` (which uses cached reader, to create a in heap slice to support  `io.ReaderAt`). But given contents are written into tmpfs file, I can simply return `os.File` disguising as `io.ReaderAt` for kernel and initrd. So there are no breaking interfaces change needed.

I have tried multiple ways of implementing lazy Opener. This version is the one I found simplest. 

On `mmap`,  If I go with breaking interface change from `io.ReaderAt` to `io.Reader` for kernel, and initrd (https://github.com/u-root/u-root/blob/main/pkg/boot/linux.go#L28), I will certainly mmap to avoid I/O. I tried it. Specifically, it would happen around here https://github.com/u-root/u-root/blob/main/pkg/boot/linux.go#L212-L233. But if the backend for `io.ReaderAt` is a file in tmpfs, more efficient way is to re-use the file, other than mmap-read it into another file. 

On `ram` consumption of a full build, I don't know if we have any existing functions that directly let us query how much physical ram our code is currently using. Probably not, we have MMU enabled, after all. But I did print out golang runtime's view of its virtual memory, and sizes of tmpfs giles. These two are, in my view, the main consumer of the ram, and they should be a good reflection.  


[1] `LazyFetch(u)` uses cached reader, which maintains an in-heap bytes slice. https://github.com/u-root/u-root/blob/main/pkg/curl/schemes.go#L216, which eventually uses a `uio.NewCachingReader(r)`, https://github.com/u-root/u-root/blob/main/pkg/curl/schemes.go#L260. If you read through the processing code, that cache was not taken advantage in case of ipxe.

Does those clarify ?

Reply:
Here, within the 'open' function, this PR does a 'io.Copy' from http body (io.Reader) to a tmpfs file (*os.File). It copies from network to tmpfs file (ram).

I could mem map the newly created tmpfs file, into a bytes slice, and io.Copy from http body (io.Reader) into the mem map slice and then flush.

On the surface, we avoid i/o. But, note that, we likely will wrap a bytes.NewBuffer on top of the mem mapped slice for writing, to make it simple. 

But as you said, it is writing to ram (tmpfs file) anyway, the speed of doing I/O write would not be penalized, as we won't be dealing with any layer of storage controller flushing to physical device. 

'io.Copy' here can arguably be replaced with mem map for the sake of reducing potential extra heap allocations it may need to do the reading. But, from my observations, golangs implemntation of io.Copy is quite lean, and I never observed large chunk of allocations lingering around, e.g while downloading > 1G payload.

Reply:
So -- I had to piece this together myself. The reason this avoids a copy is that we avoid making a new file for tmpfs just to kexec it later, which the caching reader does not do since that's in heap. Keeping it in tmpfs means you can simply kexec the file there later. That's a good point.

Please explain that succinctly in the commit description.

Can you fix the caching reader instead to do this rather than making targeted fixes that only help in specific cases? What you've done would make u-root more complicated (now there's a caching reader && this tmpfs logic, but the tmpfs logic is only in... ipxe and this catinitrd thing?)

Reply:
> Please explain that succinctly in the commit description.
Done. The commit message 1), 2), 3) did intend to say just that. I updated 3) a bit to make it clearer that we are kexec the file (re-use) in tmpfs directly. 

> Can you fix the caching reader instead to do this rather than making targeted fixes that only help in specific cases? What you've done would make u-root more complicated (now there's a caching reader && this tmpfs logic, but the tmpfs logic is only in... ipxe and this catinitrd thing?)

I thought about this. Are you sure changing the CacheReader's buffer to a `os.File` is good for all cases? - I assumed in-heap allocation is still taken advantage  in other cases.

Not sure if I agree this make u-root code more complicated. But this does add a new way, of download and cache file... It is new, and does seem complicated. But the beauty of `LazyOpenerAt` is that it takes `open func() (io.ReaderAt, error)`, which allows user to define their open function. This PR simply instructs ipxe to write a new open function.

This PR does add in new building blocks, specifically targeted for a specific case. They are building blocks, that users can opt in to use, or not to use.

**But** agree we should keep building blocks lean, and no duplicate. If there is specific suggestion on how to organize differently, happy to take them ! 

Now u-root is in **a lot of places**, I made the change to ipxe as that is where is needed for the right case. I did take a look at other netboot schemes, like FIT, they downloads, and process slightly differently that I decided to go with only ipxe.

Reply:
We can't be too shy to make changes because u-root is used in other places than here; we should just do some research before making them. https://github.com/search?q=uio.NewCachingReader&type=code -- it's only used in pkg/curl. So how about we delete it, and you make one called NewTmpfsCachingReader.

Reply:
What were the cases that [uio.NewCachingReader&type=code](https://github.com/search?q=uio.NewCachingReader&type=code) were initially written for ?

Reply:
I think what it's initially for is somewhat irrelevant -- the only thing that's relevant is what it's used for today. That search shows the only place it's used is pkg/curl, which is only AFAICT used for downloading stuff for kexecing (see [here](https://github.com/search?q=github.com%2Fu-root%2Fu-root%2Fpkg%2Fcurl&type=code#enroll-beta) -- you have to enroll the github code search beta to actually find stuff). That tells me it's safe to replace.

As for the history, not that it's relevant, but the caching reader moved from pkg/pxe to pkg/uio in [this commit](https://github.com/u-root/u-root/commit/f5e8f2eea98db116ee1ef42b8926e473b924edff). The schemes were still in pkg/pxe at the time, and that was the creation of the lazy stuff.

Reply:
I think we want to avoid regressions for one. It sounds like you sure what it was written for no longer hold for today.  Thxfor the note, Id do a check too and update here, as I recall being told the cache reader was there for some good cases, e.g. vfile